client/my-sites/patterns/index.web.tsx

```diff
@@ -4,33 +4,60 @@ import {
 	makeLayout,
 	redirectWithoutLocaleParamInFrontIfLoggedIn,
 	render as clientRender,
-} from 'calypso/controller/index.web';
+	notFound,
+} from 'calypso/controller';
 import { PatternGalleryClient } from 'calypso/my-sites/patterns/components/pattern-gallery/client';
-import { getPatternCategorySlugs } from 'calypso/my-sites/patterns/controller';
 import { PatternsHomePage } from 'calypso/my-sites/patterns/home';
-import type { RouterContext, RouterNext } from 'calypso/my-sites/patterns/types';
+import { getCurrentUserLocale } from 'calypso/state/current-user/selectors';
+import { getPatternCategoriesQueryOptions } from './hooks/use-pattern-categories';
+import type { RouterContext, RouterNext, Category } from 'calypso/my-sites/patterns/types';
 
 function renderPatterns( context: RouterContext, next: RouterNext ) {
-	context.primary = (
-		<PatternsHomePage
-			category={ context.params.category }
-			isGridView={ !! context.query.grid }
-			patternGallery={ PatternGalleryClient }
-		/>
-	);
+	if ( ! context.primary ) {
+		context.primary = (
+			<PatternsHomePage
+				category={ context.params.category }
+				isGridView={ !! context.query.grid }
+				patternGallery={ PatternGalleryClient }
+			/>
+		);
+	}
 
 	next();
 }
 
+function checkCategorySlug( context: RouterContext, next: RouterNext ) {
+	const { queryClient, lang, params, store } = context;
+	const locale = getCurrentUserLocale( store.getState() ) || lang || 'en';
+
+	// Always fetch list of categories
+	queryClient
+		.fetchQuery< Category[] >( getPatternCategoriesQueryOptions( locale ) )
+		.then( ( categories ) => {
+			// Fetch patterns only if the user is requesting a category page
+			if ( params.category ) {
+				const categoryNames = categories.map( ( category ) => category.name );
+
+				if ( ! categoryNames.includes( params.category ) ) {
+					return notFound( context, next );
```

Comment:
That's very minor but I would recommend avoiding that notation, as other developers unfamiliar with that framework may wonder what that function returns:
```suggestion
					notFound( context, next );
					
					return;
```